tools/functional-tester/etcd-runner/main.go

```diff
@@ -212,6 +215,114 @@ func runRacer(eps []string, round int) {
 	doRounds(rcs, round)
 }
 
+func runWatcher(eps []string) {
+	// watcher for setting watch on prefixes and also handling the response on watches
+	type watcher struct {
+		prefix string
+		c      *clientv3.Client
+		wc     clientv3.WatchChan
+	}
+
+	prefixes := 36       // total number of prefixes which will be watched upon
+	watchPerPrefix := 10 // number of watchers per prefix
+	ctx := context.Background()
+	prefixKeysMap := make(map[string][]string)
+	ws := make([]watcher, prefixes*watchPerPrefix)
+
+	var (
+		wg  sync.WaitGroup
+		pr  *clientv3.PutResponse
+		err error
+	)
+
+	const (
+		chars = "abcdefghijklmnopqrstuvwxyz0123456789" // for generating random keys and setting prefixes
+	)
+
+	// assign watchPerPrefix times watchers on each prefix
+	wg.Add(len(ws))
+	for i := 0; i < len(ws); {
+		for _, prefix := range chars {
+			go func(w *watcher, i int, prefix rune) {
+				defer wg.Done()
+				w.c = randClient(eps)
+				w.prefix = string(prefix)
+				w.wc = w.c.Watch(ctx, w.prefix, clientv3.WithPrefix())
+			}(&ws[i], i, prefix)
+			i++
+		}
+	}
+	wg.Wait()
+
+	done := make(chan bool)
+	// generate and put keys in cluster
+	go func(strlen int, runningTime float64, reqRate int, done chan bool) {
+		client := randClient(eps)
+		defer client.Close()
+
+		rand.Seed(time.Now().UTC().UnixNano())
+		result := make([]byte, strlen)
+		start := time.Now()
+		reqCount := 0
+		limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
+
+		for time.Since(start).Seconds() < runningTime {
+			// generate key
+			for i := 0; i < strlen; i++ {
+				result[i] = chars[rand.Intn(len(chars))]
+			}
+			key := string(result)
+
+			// limit key put as per reqRate
+			err = limiter.Wait(ctx)
+			if err != nil {
+				continue
+			}
+
+			pr, err = client.Put(ctx, key, key)
+			if err != nil {
+				log.Printf("Error occured while trying to key: %v, value : %v to kv store.", key, key)
+				continue
+			}
+
+			prefix := key[0:1]
+			vals := prefixKeysMap[prefix]
+
+			//maintain a map of key prefix and the values put for the prefix
+			prefixKeysMap[prefix] = append(vals, key)
+			reqCount++
+		}
+		done <- true
+	}(10, 30, 20, done)
+	<-done
+
+	wg.Add(len(ws))
+	for i := range ws {
+		// for all watcher verify the wtch out put with actual key updates on prefixes
+		go func(w *watcher) {
+			defer wg.Done()
+			keys, ok := prefixKeysMap[w.prefix]
+			if !ok {
+				log.Printf("No key updated on prefix : %q\n", w.prefix)
+				continue
+			}
+			for _, val := range keys {
+				watchChan := <-w.wc
```

Comment:
actually when we wait on channels and channel is not closed, it will get all available values and then expect more. But we will not get the response anymore since its expecting more values from watcher channel. And since watcher channel will not have any more values it will keep waiting on it and get stuck there.